src/Http/Routing/src/Matching/HostMatcherPolicy.cs

```diff
@@ -0,0 +1,330 @@
+            {
+                _exitDestination = exitDestination;
+                _destinations = destinations;
+            }
+
+            public override int GetDestination(HttpContext httpContext)
+            {
+                var destinations = _destinations;
+                for (var i = 0; i < destinations.Length; i++)
+                {
+                    var destination = destinations[i];
+
+                    if ((!destination.host.MatchesPort || destination.host.Port == httpContext.Request.Host.Port) &&
+                        destination.host.MatchHost(httpContext.Request.Host.Host))
+                    {
+                        return destination.destination;
+                    }
+                }
+
+                return _exitDestination;
+            }
+        }
+
+        private readonly struct EdgeKey : IEquatable<EdgeKey>, IComparable<EdgeKey>, IComparable
+        {
+            private const string WildcardHost = "*";
+            internal static readonly EdgeKey WildcardEdgeKey = new EdgeKey(null, null);
+
+            public readonly int? Port;
+            public readonly string Host;
+
+            private readonly string _wildcardEndsWith;
+
+            public EdgeKey(string host, int? port)
+            {
+                Host = host ?? WildcardHost;
+                Port = port;
+
+                HasHostWildcard = Host.StartsWith("*.", StringComparison.Ordinal);
+                _wildcardEndsWith = HasHostWildcard ? Host.Substring(1) : null;
+            }
+
+            public bool HasHostWildcard { get; }
+
+            public bool MatchesHost => !string.Equals(Host, WildcardHost, StringComparison.Ordinal);
+
+            public bool MatchesPort => Port != null;
+
+            public bool MatchesAll => !MatchesHost && !MatchesPort;
+
+            public int CompareTo(EdgeKey other)
+            {
+                var result = Comparer<string>.Default.Compare(Host, other.Host);
+                if (result != 0)
+                {
+                    return result;
+                }
+
+                return Comparer<int?>.Default.Compare(Port, other.Port);
+            }
+
+            public int CompareTo(object obj)
+            {
+                return CompareTo((EdgeKey)obj);
+            }
+
+            public bool Equals(EdgeKey other)
+            {
+                return string.Equals(Host, other.Host, StringComparison.Ordinal) && Port == other.Port;
+            }
+
+            public bool MatchHost(string host)
+            {
+                if (MatchesHost)
+                {
+                    if (HasHostWildcard)
+                    {
+                        return host.EndsWith(_wildcardEndsWith, StringComparison.OrdinalIgnoreCase);
+                    }
+                    else
+                    {
+                        return string.Equals(host, Host, StringComparison.OrdinalIgnoreCase);
+                    }
+                }
+
+                return true;
```

Comment:
@Tratcher - any special advice about this? We're adding support to routing for filtering/matching/branching on hostname+port.